app/src/batch_process.py

```diff
@@ -26,7 +26,10 @@ async def batch_process(file_path: str, engine: ChatEngineInterface) -> str:
             row.update(data)
 
         # Update fieldnames to include new columns
-        all_fieldnames = list(reader.fieldnames) + list(processed_data[0].keys())
+        all_fieldnames = list(reader.fieldnames) + [key for p in processed_data for key in p.keys()]
+        # De-duplicate while preserving order -- ignore linter warning that seen.add always returns None
+        seen = set()
+        all_fieldnames = [f for f in all_fieldnames if not (f in seen or seen.add(f))]  # type: ignore
```

Comment:
An alternative is to use a dict, which preserves order and the keys are unique. The dict values can be None.

Reply:
@yoomlam like this?

```suggestion
        all_fieldnames = {f: None for f in list(reader.fieldnames) + [key for p in processed_data for key in p.keys()]}
        all_fieldnames = list(all_fieldnames)
```

Reply:
Yes, the last line could be `all_fieldnames.keys()` to be explicit.